### PR TITLE
Art contest support and various fixes for contest system

### PR DESCRIPTION
**Pull request by nekodex**
Ongoing work on contest system, adds various fixes in addition to art contest support. Includes initial beatmap contest support but kinda incomplete.

### CODE REVIEW

File: resources/assets/coffee/react/contest.coffee [hunk 1]
```diff
@@ -16,15 +16,16 @@
...
-    hideGenre: true
-    hideBPM: true
-    hideDL: true
-    showVote: true
-    smaller: true
+    showDL: contest['type'] == 'beatmap'
+    showPreview: contest['type'] == 'music'
 
-reactTurbolinks.register 'contestTrackplayer', Trackplayer, propsFunction
+if contest['type'] == 'art'
```
Comment (nanaya) [inline resources/assets/coffee/react/contest.coffee]: not needed.

File: resources/assets/coffee/react/contest/art-list.coffee [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        waitingForResponse: @state.waitingForResponse,
+        voteCount: @state.voteCount,
+        options: @state.options,
+        contest: @state.contest
+
+    div className: 'trackplayer',
+      div className: 'trackplayer__vote-summary--art',
+        span className: 'trackplayer__vote-summary-text trackplayer__vote-summary-text--art', 'votes'
+        el Contest.VoteSummary, voteCount: @state.voteCount, maxVotes: @state.options.maxVotes
+      div style: { display: 'flex', alignItems: 'flex-start', flexWrap: 'wrap', justifyContent: 'space-around' }, entries
```
Comment (nanaya) [inline resources/assets/coffee/react/contest/art-list.coffee]: ಠ_ಠ

File: resources/views/contests/art.blade.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    See the GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+--}}
+@extends('contests.base')
+
+@section('contest-entries')
+  <div class="js-react--contestArtList"></div>
+  @include("layout.gallery_window")
```
Comment (nanaya) [inline resources/views/contests/art.blade.php]: already included in master layout?

File: resources/assets/coffee/react/contest.coffee [hunk 4]
```diff
@@ -16,15 +16,16 @@
...
+  reactTurbolinks.register 'contestArtList', Contest.ArtList, propsFunction
+
+if contest['type'] == 'beatmap' or contest['type'] == 'music'
+  reactTurbolinks.register 'contestList', Contest.EntryList, propsFunction
```
Comment (nanaya) [inline resources/assets/coffee/react/contest.coffee]: looks like both can use same main component?

**Comment by nanaya** [comment]
Seems to _work_.

File: resources/assets/coffee/react/contest/art-list.coffee [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+      el Contest.ArtEntry,
+        key: entry.id,
+        entry: entry,
+        waitingForResponse: @state.waitingForResponse,
+        voteCount: @state.voteCount,
+        options: @state.options,
+        contest: @state.contest
+
+    div className: 'trackplayer',
+      div className: 'trackplayer__vote-summary--art',
```
Comment (nanaya) [inline resources/assets/coffee/react/contest/art-list.coffee]: styles for `art` and non-`art` don't have much in common, do they? ಠ_ಠ

File: resources/assets/coffee/react/contest/entry-list.coffee [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      waitingForResponse: false
+      tracks: @props.tracks
+      voteCount: _.filter(@props.tracks, _.iteratee({selected: true})).length
+      contest: @props.contest
+      options:
+        showDL: @props.options.showDL ? false
+        showPreview: @props.options.showPreview ? false
+        maxVotes: @props.contest.max_votes ? 3
+
+  handleVoteClick: (_e, {track_id, callback}) =>
```
Comment (nanaya) [inline resources/assets/coffee/react/contest/entry-list.coffee]: \*scrolls up\*
\*scrolls down\*

hmmmmm

(also why one called "EntryList" and the other "ArtList")

**Comment by peppy** [comment]
merging this for now so we can make contest happen. fixes can come after 👍

**Pull request merged**